### PR TITLE
Enable linking for UCR Expressions 

### PR DESCRIPTION
**Pull request by proteusvacuum**
## Product Description

Adds UCR Registry Expressions and Filters to linked domains. Currently remote links are not supported, but we can use the same pattern as linking reports remotely for this, although for now this isn't required.

## Technical Summary

https://dimagi-dev.atlassian.net/browse/SC-2008

## Feature Flag

UCR_REGISTRY, LINKED_DOMAINS

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/linked_domain/templates/linked_domain/partials/push_content.html
Comment (gherceg) [inline corehq/apps/linked_domain/templates/linked_domain/partials/push_content.html:88]: Is this a scenario that gets hit when dealing with an upstream domain that has only remote downstream domains?

File: corehq/apps/linked_domain/view_helpers.py
Comment (gherceg) [inline corehq/apps/linked_domain/view_helpers.py:193]: There are some tests in `linked_domain/tests/test_view_helpers.py` that test build view model methods. I'll admit it feels a bit tedious, but is also a quick and easy way to ensure view models are built properly.

File: corehq/apps/linked_domain/view_helpers.py
Comment (gherceg) [inline corehq/apps/linked_domain/view_helpers.py:303]: This data model is feature flag dependent right? I see that it doesn't fit gracefully into the current design which assumes all feature flag data models are at the "domain" level. That being said, there should still be a check that the feature flag is enabled prior to building/appending these view models right? Definitely an edge case, but possible.

File: corehq/apps/linked_domain/tasks.py
Comment (gherceg) [inline corehq/apps/linked_domain/tasks.py:211]: 🙌🏻

File: corehq/apps/linked_domain/view_helpers.py
Comment (gherceg) [inline corehq/apps/linked_domain/view_helpers.py:118]: I'm regretting my liberal use of the word "list" in these docstrings, but don't expect any changes here.

File: corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py
Comment (gherceg) [inline corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py:6]: nit: we've been moving towards isolating migrations when possible just to decrease the chances of dealing with a migration if the need to rollback a PR arises.

**Review by gherceg — COMMENTED** [review]
Looks great! Holding off on approval until some questions are answered.

File: corehq/apps/linked_domain/view_helpers.py
Comment (sravfeyn) [inline corehq/apps/linked_domain/view_helpers.py:361]: I think this can be simplified with (which also avoids a nested try-catch)
```
ucr_expression  = ucr_expressions.pop(ucr_expression_id, None)
```

File: corehq/apps/linked_domain/view_helpers.py
Comment (proteusvacuum) [reply corehq/apps/linked_domain/view_helpers.py:303]: Yes, this data model is FF dependent. I explicitly didn't check if the FF is enabled on the upstream domain since this will only add items into the view if there are any that exist for the domain in question (which would have required the FF to be enabled at some point). I'm always conflicted about how often we should check for a feature flag - removing a domain from this flag wouldn't delete the objects from the DB, so I feel like linking these items is still a valid operation. Happy to defer to you if you feel otherwise though.

File: corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py
Comment (proteusvacuum) [reply corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py:6]: Interesting - by isolating, do you mean I should spin this migration out into a separate PR, or just a separate commit?

File: corehq/apps/linked_domain/view_helpers.py
Comment (proteusvacuum) [reply corehq/apps/linked_domain/view_helpers.py:361]: You are right! I just copied the way this was done for the other items in this file and didn't think too hard about it. I will update.

File: corehq/apps/linked_domain/templates/linked_domain/partials/push_content.html
Comment (proteusvacuum) [reply corehq/apps/linked_domain/templates/linked_domain/partials/push_content.html:88]: Yes. I think in most cases with remote downstreams, there aren't any local downstreams too.

File: corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py
Comment (gherceg) [reply corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py:6]: Yeah separating the migration into its own PR, though now I'm rethinking this. We don't appear to have this written in documentation anywhere, and I can see reasons to keep code and migration together. So it feels more like a case-by-case basis and I think this is fine to have them together given that you are only adding new functionality.

File: corehq/apps/linked_domain/view_helpers.py
Comment (gherceg) [reply corehq/apps/linked_domain/view_helpers.py:303]: > removing a domain from this flag wouldn't delete the objects from the DB, so I feel like linking these items is still a valid operation

This is the edge case I was thinking of. My assumption was that once a domain is removed from a feature flag, it loses access to anything related to that feature flag. That's just an assumption though, and you know this flag and its intent better than I do so deferring to you 👍🏻

File: corehq/apps/linked_domain/view_helpers.py
Comment (proteusvacuum) [reply corehq/apps/linked_domain/view_helpers.py:193]: I added more tests there in the same style as the others.

File: corehq/apps/linked_domain/view_helpers.py
Comment (proteusvacuum) [reply corehq/apps/linked_domain/view_helpers.py]: https://github.com/dimagi/commcare-hq/pull/31597/commits/55dba1503f7d3ddc02634044a248a0391286986e

File: corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py
Comment (proteusvacuum) [reply corehq/apps/userreports/migrations/0019_ucrexpression_upstream_id.py:6]: Yeah, I can understand splitting out a migration that would migrate _data_ so it can be run separately. Migrations like that are likely hard to revert.

File: corehq/apps/linked_domain/tests/test_view_helpers.py
Comment (gherceg) [inline corehq/apps/linked_domain/tests/test_view_helpers.py:476]: Assume this accidentally made its way into the commit.

**Review by gherceg — APPROVED** [review]
Nothing blocking

File: corehq/apps/linked_domain/tests/test_view_helpers.py
Comment (proteusvacuum) [reply corehq/apps/linked_domain/tests/test_view_helpers.py]: 💯 my bad.

**Pull request merged**
